### PR TITLE
Backprop wrt candidates in gen_candidates_torch

### PR DESCRIPTION
**Pull request by wjmaddox**
## Motivation

Bugfix for #765. The current error was occurring when trying to optimize qKnowledgeGradient with gen_candidates_torch. The fix is to backprop only wrt candidates like what occurs in `gen_candidates_scipy`. 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Not sure how to setup a unit test here.

## Related PRs

N/A.

### CODE REVIEW

**Comment by saitcakmak** [comment]
> Not sure how to setup a unit test here.

In #765 the bug was in optimizing qKG. The current test suite only optimizes qEI (thus missing the issue). So, adding qKG as an additional test case may be an option

File: botorch/generation/gen.py [hunk 1]
```diff
@@ -219,11 +219,12 @@ def gen_candidates_torch(
...
             print("Iter: {} - Value: {:.3f}".format(i, -(loss.item())))
         loss_trajectory.append(loss.item())
-        param_trajectory["candidates"].append(candidates.clone())
+        param_trajectory["candidates"].append(candidates.clone().detach())
 
         def closure():
             bayes_optimizer.zero_grad()
-            loss.backward()
+            output_grad = torch.autograd.grad(loss, candidates)[0]
+            candidates.grad = output_grad
```
Comment (Balandat) [inline botorch/generation/gen.py:227]: hmm there seems to be some inconsistency here between the unconstrained candidates, the clamped candidates, and those were `fixed_features` is applied. Specifically, `bayes_optimizer` uses `clamped_candidates` as its parameter (l. 207), but here we're taking the gradient w.r.t. `candidates`. There also seems to be some inconsistency below where we compute `clamped_candidates` but return `candidates` (and evaluate the acquisition function on those). 

We probably want to combine the clamping and the "fixing" of features into a single wrapper so we can always work with a single `projected_candidates` tensor to avoid this kind of confusion.

**Comment by Balandat** [comment]
Is this issue addressed by #839? If not maybe @vishwakftw can grab the relevant changes and incorporate them?

**Comment by wjmaddox** [comment]
I don't know, probably the inconsistency you pointed out above is. Will checkout #839 now and see if that resolves the issue.

**Comment by wjmaddox** [comment]
Re-running the command in #765 on top of #839 still fails so this should probably stay open. Maybe @vishwakftw can resolve the second gen conversation here since I don't have a great idea of what should be fixed / transformed?

File: botorch/generation/gen.py [hunk 2]
```diff
@@ -288,19 +286,18 @@ def gen_candidates_torch(
...
+            output_grad = torch.autograd.grad(loss, clamped_candidates)[0]
+            clamped_candidates.grad = output_grad
             return loss
 
         bayes_optimizer.step(closure)
         with torch.no_grad():
             clamped_candidates = columnwise_clamp(
                 X=clamped_candidates, lower=lower_bounds, upper=upper_bounds
-            )
+            ).requires_grad_(True)
```
Comment (Balandat) [inline botorch/generation/gen.py:300]: Hmm I've never thought about what happens when you set `requires_grad` within a `no_grad` context. Will this work properly?

File: test/generation/test_gen.py [hunk 3]
```diff
@@ -72,15 +72,28 @@ def test_gen_candidates(self, gen_candidates=gen_candidates_scipy, options=None)
...
+
+                candidates, _ = gen_candidates(
+                    initial_conditions=ics,
+                    acquisition_function=acqf,
+                    lower_bounds=0,
+                    upper_bounds=1,
+                    options=options or {},
+                )
+                if isinstance(acqf, qKnowledgeGradient):
+                    candidates = candidates[0]
```
Comment (Balandat) [inline test/generation/test_gen.py:94]:
```suggestion
                if isinstance(acqf, qKnowledgeGradient):
                    candidates = acqf.extract_candidates(candidates)
```

File: botorch/generation/gen.py [hunk 4]
```diff
@@ -288,19 +286,18 @@ def gen_candidates_torch(
         loss = -acquisition_function(clamped_candidates).sum()
         if verbose:
             print("Iter: {} - Value: {:.3f}".format(i, -(loss.item())))
-        loss_trajectory.append(loss.item())
-        param_trajectory["candidates"].append(clamped_candidates.clone())
```
Comment (Balandat) [inline botorch/generation/gen.py:292]: I guess this code has been dead for a while then?

File: test/generation/test_gen.py [hunk 5]
```diff
@@ -96,18 +109,30 @@ def test_gen_candidates_with_none_fixed_features(
...
+                candidates, _ = gen_candidates(
+                    initial_conditions=ics,
+                    acquisition_function=acqf,
+                    lower_bounds=0,
+                    upper_bounds=1,
+                    fixed_features={1: None},
+                    options=options or {},
+                )
+                if isinstance(acqf, qKnowledgeGradient):
+                    candidates = candidates[0]
```
Comment (Balandat) [inline test/generation/test_gen.py:132]:
```suggestion
                    candidates = acqf.extract_candidates(candidates)
```

File: test/generation/test_gen.py [hunk 6]
```diff
@@ -121,18 +146,32 @@ def test_gen_candidates_with_fixed_features(
...
+                    initial_conditions=ics,
+                    acquisition_function=acqf,
+                    lower_bounds=0,
+                    upper_bounds=1,
+                    fixed_features={1: 0.25},
+                    options=options,
+                )
+
+                if isinstance(acqf, qKnowledgeGradient):
+                    candidates = candidates[0]
```
Comment (Balandat) [inline test/generation/test_gen.py:170]:
```suggestion
                    candidates = acqf.extract_candidates(candidates)
```

**Review by Balandat — APPROVED** [review]

File: botorch/generation/gen.py (hunk 4 above)
Comment (wjmaddox) [reply botorch/generation/gen.py:292]: Yes.

File: botorch/generation/gen.py (hunk 2 above)
Comment (wjmaddox) [reply botorch/generation/gen.py:300]: Seems to be since it's not actually a differentiable operation. Wasn't passing the unit tests before.
